server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java

```diff
@@ -171,11 +171,18 @@ public Storage selectStorage(ShuffleDataFlushEvent event) {
             partitionId
         )
     );
-    event.setUnderStorage(storage);
 
     // store it to cache.
-    partitionsOfStorage.put(UnionKey.toKey(appId, shuffleId, partitionId), storage);
-    return storage;
+    final LocalStorage selectedStorage = storage;
+    return partitionsOfStorage.compute(
```

Comment:
Use the concurrent hashmap compute method to ensure the thread safe

cc @advancedxy , please help check this part.